### PR TITLE
nix: patch curl netrc regression

### LINKED ISSUE #406: `netrc` omits `login`, defaulting to system username
Because the `netrc` configuration does not specify a value for `login` ([here](https://github.com/cachix/cachix/blob/0295dac48d2857f497cf3a9b6467f399853743b7/cachix/src/Cachix/Client/NetRc.hs#L51)), it will use the fallback of the current username in the system.

This will cause an authentication error when using a cache where your Cachix username is not the same as your system username on the system you are trying to connect to Cachix with. This was an issue for me because my Cachix username comes from my GitHub account, and my system username is of course allowed to be whatever I like.

### PR DESCRIPTION
**Pull request by domenkozar**
Upstream bug https://github.com/curl/curl/issues/8653

Fixes https://github.com/cachix/cachix/issues/406

This is a hack to not trigger stdenv rebuild but fix Nix in the fastest manner.

Following this merge I'll backport to 22.05 and make a stdenv rebuild version for staging-next.

### CODE REVIEW

**Comment by domenkozar** [comment]
I'm going to merge this by the end of the day (US time) as this is impacting a lot of Cachix users.

**Review by Mindavi — APPROVED** [review]
Ok for now, I guess. <s>And only affects cachix users if it turns out to not be correct/complete or unsafe.</s>

File: pkgs/tools/networking/curl/netrc-regression.patch
Comment (roberth) [inline pkgs/tools/networking/curl/netrc-regression.patch:10]: If I read this correctly, a blank user will also be set if the password is blank. That's not necessary for cachix and it's slightly risky.
Does curl come with a test that exercises the usual user+pass both set combination, to make sure we don't accidentally clear the user? And do we run it?

**Comment by roberth** [comment]
> And only affects cachix users if it turns out to not be correct/complete or unsafe.

This is not the case. It affects all nix users.

**Comment by roberth** [comment]
Did you create a PR upstream? It'd be good to have their feedback (also after merge).
If you could add a comment with a link to it, that will be helpful for maintainers to track the change.

File: pkgs/tools/networking/curl/netrc-regression.patch
Comment (domenkozar) [reply pkgs/tools/networking/curl/netrc-regression.patch:10]: It comes with a test suite, but it doesn't seem to have caught this bug. And yes, we run the tests.

About the empty password case, do you have a suggestion how to handle it better?

Comment (roberth) [reply pkgs/tools/networking/curl/netrc-regression.patch:10]: I've misread it, but let's be safe and check the pointer before checking for a blank string.
```suggestion
+    if(userp && !*userp && passwdp && *passwdp) {
```

ie make the blank check null-pointer safe.

Honestly C code with this much pointers makes me very anxious. I have a lot of respect for people who can program in such an environment.

**Comment by roberth** [comment]
Could you add a link to the issue in a comment?

The author seems to be aware of demand for his patch and has been asked to make a PR. Maybe you could ask him again, or offer to credit him in the commit's author field if he's ok with that.

**Comment by domenkozar** [comment]
> Could you add a link to the issue in a comment?
> 
> The author seems to be aware of demand for his patch and has been asked to make a PR. Maybe you could ask him again, or offer to credit him in the commit's author field if he's ok with that.

I plan to make a follow up PR that applies the patch unconditionally to staging-next, so this code will live here for no more than a few days.

**Comment by roberth** [comment]
Fair enough. Better to focus your attention on working with upstream then. We don't want to have a situation like debian had with their openssl entropy bug, existing only in a distro patch.

**Comment by zowoq** [comment]
https://github.com/curl/curl/commit/8bd03516d62ab08972190e2ccdaa06ab1d692125

Seems this is fixed in 7.85.0 so I'll remove it as part of https://github.com/NixOS/nixpkgs/pull/189083

**Pull request merged**
